File: AgapeNFS-e/_legacy/python/src/agape_nfse/parser.py
from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass
from html import unescape

from bs4 import BeautifulSoup, Tag

from .payloads import FilterFields


@dataclass(frozen=True)
class NotaFiscal:
    codigo: str
    numero: str
    exercicio: str
    contribuinte: str
    destinatario: str
    emissao: str
    total: str
    xml_action_name: str

    @property
    def key(self) -> str:
        return f"{self.exercicio}-{self.numero}-{self.codigo}"


@dataclass(frozen=True)
class NotaPage:
    table_id: str
    current_page: int
    page_links: set[int]
    has_next: bool
    notas: list[NotaFiscal]
    view_state: str


def soup_from_html(html: str | bytes) -> BeautifulSoup:
    return BeautifulSoup(html, "html.parser")


def parse_view_state(html: str | bytes) -> str:
    soup = soup_from_html(html)
    field = soup.find("input", attrs={"name": "javax.faces.ViewState"})
    if not field:
        raise ValueError("Nao foi possivel localizar javax.faces.ViewState.")
    value = str(field.get("value") or "").strip()
    if not value:
        raise ValueError("javax.faces.ViewState encontrado sem valor.")
    return value


def parse_filter_fields(html: str | bytes, view_state: str | None = None) -> FilterFields:
    soup = soup_from_html(html)
    form = soup.find("form", id="form_conteudo") or soup.find("form", attrs={"name": "form_conteudo"})
    if not isinstance(form, Tag):
        raise ValueError("Formulario form_conteudo nao encontrado.")

    date_inputs = [
        inp
        for inp in form.find_all("input")
        if _attr(inp, "type") == "text" and "rich-calendar-input" in " ".join(inp.get("class", []))
    ]
    if len(date_inputs) < 2:
        raise ValueError("Campos de data do filtro nao encontrados.")

    pesquisa = _name_or_fail(form.find("input", attrs={"name": re.compile(r":pesquisa$")}), "campo pesquisa")
    start_date = _name_or_fail(date_inputs[0], "data inicial")
    end_date = _name_or_fail(date_inputs[1], "data final")
    start_current = _name_or_fail(form.find("input", attrs={"name": f"{start_date.replace('InputDate', 'InputCurrentDate')}"}), "mes atual inicial")
    end_current = _name_or_fail(form.find("input", attrs={"name": f"{end_date.replace('InputDate', 'InputCurrentDate')}"}), "mes atual final")

    text_inputs = [
        inp
        for inp in form.find_all("input")
        if _attr(inp, "type") == "text"
        and _name(inp) not in {pesquisa, start_date, end_date}
        and "rich-calendar-input" not in " ".join(inp.get("class", []))
    ]
    if not text_inputs:
        raise ValueError("Campo de exercicio nao encontrado.")
    exercise = _name_or_fail(text_inputs[0], "exercicio")

    status = _name_or_fail(form.find("input", attrs={"type": "radio", "value": "F"}), "status finalizada")
    form_marker = _name_or_fail(form.find("input", attrs={"name": "form_conteudo"}), "marcador do formulario")
    auto_scroll = _name_or_fail(form.find("input", attrs={"name": "autoScroll"}), "autoScroll")
    submit = form.find("a", id=re.compile(r":lupa$")) or form.find("a", attrs={"name": re.compile(r":lupa$")})
    submit_name = _name_or_fail(submit, "botao de pesquisa")

    return FilterFields(
        pesquisa=pesquisa,
        start_date=start_date,
        start_current_month=start_current,
        end_date=end_date,
        end_current_month=end_current,
        exercise=exercise,
        status=status,
        form_marker=form_marker,
        auto_scroll=auto_scroll,
        view_state=view_state or parse_view_state(html),
        submit_name=submit_name,
    )


def parse_nota_page(html: str | bytes, fallback_view_state: str = "") -> NotaPage:
    soup = soup_from_html(html)
    try:
        table = _find_notes_table(soup)
    except ValueError:
        view_state = fallback_view_state
        try:
            view_state = parse_view_state(html)
        except ValueError:
            pass
        return NotaPage(table_id="", current_page=1, page_links={1}, has_next=False, notas=[], view_state=view_state)
    table_id = str(table.get("id") or "")
    columns = _header_columns(table)

    required = {
        "codigo": _find_header(columns, "codigo"),
        "numero": _find_header(columns, "numero"),
        "exercicio": _find_header(columns, "exercicio"),
        "contribuinte": _find_header(columns, "contribuinte"),
        "destinatario": _find_header(columns, "destinatario"),
        "emissao": _find_header(columns, "data de emissao"),
        "total": _find_header(columns, "total"),
        "xml": _find_header(columns, "xml"),
    }

    tbody = table.find("tbody", id=re.compile(r":tb$")) or table.find("tbody")
    rows = tbody.find_all("tr", recursive=False) if isinstance(tbody, Tag) else []
    notas: list[NotaFiscal] = []
    for row in rows:
        xml_cell = _cell_by_suffix(row, required["xml"])
        action_name = _extract_xml_action_name(xml_cell)
        if not action_name:
            continue
        notas.append(
            NotaFiscal(
                codigo=_cell_text(row, required["codigo"]),
                numero=_cell_text(row, required["numero"]),
                exercicio=_cell_text(row, required["exercicio"]),
                contribuinte=_cell_text(row, required["contribuinte"]),
                destinatario=_cell_text(row, required["destinatario"]),
                emissao=_cell_text(row, required["emissao"]),
                total=_cell_text(row, required["total"]),
                xml_action_name=action_name,
            )
        )

    view_state = fallback_view_state
    try:
        view_state = parse_view_state(html)
    except ValueError:
        pass

    return NotaPage(
        table_id=table_id,
        current_page=_parse_current_page(table),
        page_links=_parse_page_links(table),
        has_next=_parse_has_next(table),
        notas=notas,
        view_state=view_state,
    )


def _find_notes_table(soup: BeautifulSoup) -> Tag:
    for table in soup.find_all("table"):
        headers = [_normalize_text(th.get_text(" ", strip=True)) for th in table.find_all("th")]
        if "numero" in headers and "xml" in headers and "data de emissao" in headers:
            return table
    raise ValueError("Tabela de notas nao encontrada.")


def _header_columns(table: Tag) -> dict[str, str]:
    columns: dict[str, str] = {}
    for th in table.find_all("th"):
        text = _normalize_text(th.get_text(" ", strip=True))
        th_id = str(th.get("id") or "")
        match = re.search(r":(j_id\d+)header$", th_id)
        if text and match:
            columns[text] = match.group(1)
    return columns


def _find_header(columns: dict[str, str], header: str) -> str:
    normalized = _normalize_text(header)
    if normalized in columns:
        return columns[normalized]
    raise ValueError(f"Coluna obrigatoria nao encontrada: {header}.")


def _cell_by_suffix(row: Tag, suffix: str) -> Tag | None:
    for cell in row.find_all("td", recursive=False):
        if str(cell.get("id") or "").endswith(f":{suffix}"):
            return cell
    return None


def _cell_text(row: Tag, suffix: str) -> str:
    cell = _cell_by_suffix(row, suffix)
    return cell.get_text(" ", strip=True) if isinstance(cell, Tag) else ""


def _extract_xml_action_name(cell: Tag | None) -> str:
    if not isinstance(cell, Tag):
        return ""
    link = cell.find("a", onclick=True)
    if not isinstance(link, Tag):
        return ""
    onclick = unescape(str(link.get("onclick") or ""))
    match = re.search(r"jsfcljs\(document\.forms\['form_conteudo'\],'([^,']+),\1'", onclick)
    return match.group(1) if match else ""


def _parse_current_page(table: Tag) -> int:
    active = table.find(class_=lambda value: value and "rich-datascr-act" in str(value))
    if not isinstance(active, Tag):
        return 1
    text = active.get_text(strip=True)
    return int(text) if text.isdigit() else 1


def _parse_page_links(table: Tag) -> set[int]:
    pages: set[int] = set()
    for cell in table.find_all(class_=lambda value: value and "rich-datascr" in str(value)):
        text = cell.get_text(strip=True)
        if text.isdigit():
            pages.add(int(text))
    active = _parse_current_page(table)
    pages.add(active)
    return pages


def _parse_has_next(table: Tag) -> bool:
    for cell in table.find_all(onclick=True):
        onclick = str(cell.get("onclick") or "")
        classes = " ".join(cell.get("class", []))
        if "'page': 'next'" in onclick and "rich-datascr-button-dsbld" not in classes:
            return True
    return False


def _normalize_text(value: str) -> str:
    normalized = unicodedata.normalize("NFKD", value)
    ascii_text = normalized.encode("ascii", "ignore").decode("ascii")
    return re.sub(r"\s+", " ", ascii_text).strip().lower()


def _name(tag: Tag | None) -> str:
    if not isinstance(tag, Tag):
        return ""
    return str(tag.get("name") or tag.get("id") or "")


def _name_or_fail(tag: Tag | None, label: str) -> str:
    name = _name(tag)
    if not name:
        raise ValueError(f"Nao foi possivel localizar {label}.")
    return name


def _attr(tag: Tag, name: str) -> str:
    return str(tag.get(name) or "").lower()

File: AgapeNFS-e/src/agape/parser.ts
import * as cheerio from "cheerio";
import type { AnyNode } from "domhandler";
import type { Cheerio, CheerioAPI } from "cheerio";
import type { FilterFields, NotaFiscal, NotaPage } from "./types";

type NodeSet = Cheerio<AnyNode>;

export function parseViewState(html: string | Buffer): string {
  const $ = cheerio.load(html.toString());
  const value = $('input[name="javax.faces.ViewState"]').first().attr("value")?.trim() ?? "";
  if (!value) {
    throw new Error("Nao foi possivel localizar javax.faces.ViewState.");
  }
  return value;
}

export function parseFilterFields(html: string | Buffer, viewState?: string): FilterFields {
  const $ = cheerio.load(html.toString());
  const form = $("#form_conteudo").first().length
    ? $("#form_conteudo").first()
    : $('form[name="form_conteudo"]').first();
  if (!form.length) {
    throw new Error("Formulario form_conteudo nao encontrado.");
  }

  const dateInputs = form
    .find("input")
    .filter((_, el) => attr($(el), "type") === "text" && classes($(el)).includes("rich-calendar-input"))
    .toArray();
  if (dateInputs.length < 2) {
    throw new Error("Campos de data do filtro nao encontrados.");
  }

  const pesquisa = nameOrFail(
    form.find("input").filter((_, el) => name($(el)).endsWith(":pesquisa")).first(),
    "campo pesquisa",
  );
  const startDate = nameOrFail($(dateInputs[0]), "data inicial");
  const endDate = nameOrFail($(dateInputs[1]), "data final");
  const startCurrentMonth = nameOrFail(findByName($, form, startDate.replace("InputDate", "InputCurrentDate")), "mes atual inicial");
  const endCurrentMonth = nameOrFail(findByName($, form, endDate.replace("InputDate", "InputCurrentDate")), "mes atual final");

  const textInputs = form
    .find("input")
    .filter((_, el) => {
      const input = $(el);
      const inputName = name(input);
      return (
        attr(input, "type") === "text" &&
        inputName !== pesquisa &&
        inputName !== startDate &&
        inputName !== endDate &&
        !classes(input).includes("rich-calendar-input")
      );
    });
  if (!textInputs.length) {
    throw new Error("Campo de exercicio nao encontrado.");
  }

  const status = nameOrFail(form.find('input[type="radio"][value="F"]').first(), "status finalizada");
  const formMarker = nameOrFail(form.find('input[name="form_conteudo"]').first(), "marcador do formulario");
  const autoScroll = nameOrFail(form.find('input[name="autoScroll"]').first(), "autoScroll");
  const submit = form
    .find("a")
    .filter((_, el) => {
      const link = $(el);
      return name(link).endsWith(":lupa") || (link.attr("id") ?? "").endsWith(":lupa");
    })
    .first();
  const submitName = nameOrFail(submit, "botao de pesquisa");

  return {
    pesquisa,
    startDate,
    startCurrentMonth,
    endDate,
    endCurrentMonth,
    exercise: nameOrFail(textInputs.first(), "exercicio"),
    status,
    formMarker,
    autoScroll,
    viewState: viewState ?? parseViewState(html),
    submitName,
  };
}

export function parseNotaPage(html: string | Buffer, fallbackViewState = ""): NotaPage {
  const $ = cheerio.load(html.toString());
  const table = findNotesTable($);
  if (!table.length) {
    return {
      tableId: "",
      currentPage: 1,
      pageLinks: new Set([1]),
      hasNext: false,
      notas: [],
      viewState: readViewStateOrFallback(html, fallbackViewState),
    };
  }

  const tableId = table.attr("id") ?? "";
  const columns = headerColumns($, table);
  const required = {
    codigo: findHeader(columns, "codigo"),
    numero: findHeader(columns, "numero"),
    exercicio: findHeader(columns, "exercicio"),
    contribuinte: findHeader(columns, "contribuinte"),
    destinatario: findHeader(columns, "destinatario"),
    emissao: findHeader(columns, "data de emissao"),
    total: findHeader(columns, "total"),
    xml: findHeader(columns, "xml"),
  };

  const tbody = table.find("tbody").filter((_, el) => (($(el).attr("id") ?? "").endsWith(":tb"))).first();
  const rows = (tbody.length ? tbody : table.find("tbody").first()).children("tr");
  const notas: NotaFiscal[] = [];

  rows.each((_, rowEl) => {
    const row = $(rowEl);
    const xmlCell = cellBySuffix($, row, required.xml);
    const xmlActionName = extractXmlActionName($, xmlCell);
    if (!xmlActionName) {
      return;
    }

    notas.push({
      codigo: cellText($, row, required.codigo),
      numero: cellText($, row, required.numero),
      exercicio: cellText($, row, required.exercicio),
      contribuinte: cellText($, row, required.contribuinte),
      destinatario: cellText($, row, required.destinatario),
      emissao: cellText($, row, required.emissao),
      total: cellText($, row, required.total),
      xmlActionName,
    });
  });

  return {
    tableId,
    currentPage: parseCurrentPage($, table),
    pageLinks: parsePageLinks($, table),
    hasNext: parseHasNext($, table),
    notas,
    viewState: readViewStateOrFallback(html, fallbackViewState),
  };
}

function findNotesTable($: CheerioAPI): NodeSet {
  const tables = $("table").toArray();
  const found = tables.find((table) => {
    const headers = $(table)
      .find("th")
      .map((_, th) => normalizeText($(th).text()))
      .toArray();
    return headers.includes("numero") && headers.includes("xml") && headers.includes("data de emissao");
  });
  return found ? $(found) : $();
}

function headerColumns($: CheerioAPI, table: NodeSet): Record<string, string> {
  const columns: Record<string, string> = {};
  table.find("th").each((_, th) => {
    const text = normalizeText($(th).text());
    const thId = $(th).attr("id") ?? "";
    const match = /:(j_id\d+)header$/.exec(thId);
    if (text && match?.[1]) {
      columns[text] = match[1];
    }
  });
  return columns;
}

function findHeader(columns: Record<string, string>, header: string): string {
  const normalized = normalizeText(header);
  const value = columns[normalized];
  if (!value) {
    throw new Error(`Coluna obrigatoria nao encontrada: ${header}.`);
  }
  return value;
}

function cellBySuffix($: CheerioAPI, row: NodeSet, suffix: string): NodeSet {
  const found = row.children("td").toArray().find((cell) => (($(cell).attr("id") ?? "").endsWith(`:${suffix}`)));
  return found ? $(found) : $();
}

function cellText($: CheerioAPI, row: NodeSet, suffix: string): string {
  return cellBySuffix($, row, suffix).text().replace(/\s+/g, " ").trim();
}

function extractXmlActionName($: CheerioAPI, cell: NodeSet): string {
  const onclick = cell.find("a[onclick]").first().attr("onclick") ?? "";
  const match = /jsfcljs\(document\.forms\['form_conteudo'\],'([^,']+),\1'/.exec(onclick);
  return match?.[1] ?? "";
}

function parseCurrentPage($: CheerioAPI, table: NodeSet): number {
  const active = table.find(".rich-datascr-act").first().text().trim();
  return /^\d+$/.test(active) ? Number(active) : 1;
}

function parsePageLinks($: CheerioAPI, table: NodeSet): Set<number> {
  const pages = new Set<number>();
  table.find('[class*="rich-datascr"]').each((_, cell) => {
    const text = $(cell).text().trim();
    if (/^\d+$/.test(text)) {
      pages.add(Number(text));
    }
  });
  pages.add(parseCurrentPage($, table));
  return pages;
}

function parseHasNext($: CheerioAPI, table: NodeSet): boolean {
  return table
    .find("[onclick]")
    .toArray()
    .some((cell) => {
      const item = $(cell);
      return (item.attr("onclick") ?? "").includes("'page': 'next'") && !classes(item).includes("rich-datascr-button-dsbld");
    });
}

function readViewStateOrFallback(html: string | Buffer, fallback: string): string {
  try {
    return parseViewState(html);
  } catch {
    return fallback;
  }
}

function normalizeText(value: string): string {
  return value
    .normalize("NFD")
    .replace(/\p{Diacritic}/gu, "")
    .replace(/\s+/g, " ")
    .trim()
    .toLowerCase();
}

function findByName($: CheerioAPI, parent: NodeSet, expectedName: string): NodeSet {
  return parent.find("input").filter((_, el) => name($(el)) === expectedName).first();
}

function name(tag: NodeSet): string {
  return tag.attr("name") ?? tag.attr("id") ?? "";
}

function nameOrFail(tag: NodeSet, label: string): string {
  const value = name(tag);
  if (!value) {
    throw new Error(`Nao foi possivel localizar ${label}.`);
  }
  return value;
}

function attr(tag: NodeSet, attrName: string): string {
  return (tag.attr(attrName) ?? "").toLowerCase();
}

function classes(tag: NodeSet): string {
  return tag.attr("class") ?? "";
}
